### ISSUE TITLE
Add caching to gitlab-ci example

### ISSUE THREAD

**Issue by ochorocho** [ochorocho/ddev-docker#4]
Caching the downloaded docker images on a project base would make sense to speed up builds.
Currently, all images are download on each and every job run.

**Comment by rfay** [comment]
Our experience with trying to cache images in GitHub hasn't been very successful, largely because just saving them or extracting them takes more time than downloading them in the first place.

**Comment by ayalon** [comment]
@rfay @ochorocho 
I was thinking a lot about how we could cache the pipeline stuff.
For the moment, this is the biggest draw back of this approach.

I was thinking of a possible solution.

**Is DDEV somehow capable of building and pushing a docker image to a registry?**

If this would be possible, we could somehow check if the DDEV configuration has changed (maybe a salt.txt file or similar). 
In the pipeline we start and build a DDEV container first and push the resulting docker image to the registry like that:
`docker push $CI_REGISTRY_IMAGE:$DDEV_SALT`

If the DDEV configuration did not change, we can just pull the image from the registry and run the ddev tests.

Or are there other ideas how to avoid the constant download of all docker layers within the container?

**Comment by rfay** [comment]
DDEV's docker images are all in the hub.docker.com registry. That's the problem here, we don't have a way to efficiently store the images locally. Downloading them doesn't take long, but unpacking/extracting *does* take a long time. So caching locally (which isn't hard) still takes a long time because the extraction into the local (fresh) docker instance does take time even if the download is fast.

**Comment by ayalon** [comment]
But isn't DDEV building a new docker image with `docker compose`?

Locally I can profit from the fact, that the layer are cached. I understand that downloading is as fast as geting them from the cache.

But I thought We could put the resulting image from docker compose which uses the DDEV docker images as a base could be pushed to a registry and reused.

Or is there another mechanism possibly speed up the process? Instead of running Docker-in-Docker, we build the DDEV final docker image, then push it to gitlab registry and in the next step we run the tests inside this prebuild DDEV image.

Maybe this is also the wrong approach. Do you have another idea? We are using kaniko to build and push docker images and I was imaging that this could be a way to save some resources in the pipeline.

**Comment by rfay** [comment]
DDEV adds a new layer to the image at start, but the key problem with testing is the (download and) extraction of the base image, especially ddev/ddev-webserver. The addition of the extra layers (for username, etc.) takes nearly no time. You can watch it yourself on a `ddev start`. The dots show the build time.

IMO the general problem is figuring out how to have the actual docker server persist state, so that images are ready there when they're needed.

**Comment by ayalon** [comment]
Ok I understand. You are absolutely right. Unfortunatly I lack the knowledge how to do that. I'm not even sure that it is technically possible.

But I agree that if the base image would be available, the process would be very fast.

**Comment by ochorocho** [comment]
If the images are stored in ddev itself and not in the DIND-Service image we could certainly extend the image to contain the images.
Just dunno if these images vary depending on the service version in use.

I'd like to keep this image small if possible.

https://github.com/ddev/ddev/issues/4479

**Comment by rfay** [comment]
You point to 
* https://github.com/ddev/ddev/issues/4479

for some reason... but that was fixed ages ago so `ddev debug download-images` does not require a project.

**Comment by ochorocho** [comment]
👍 my intention was to use `ddev debug download-images` to download the images and ship this image with the images pre-downloaded.

**Comment by rfay** [comment]
Will be great to see how that comes out!

**Comment by ayalon** [comment]
Great idea! That make things faster for sure!

**Comment by AkibaAT** [comment]
I've had pretty decent success improving the pipeline duration by building the required images only once, and pushing them to the registry for reuse, as well as using the GitLab dependency proxy to pull in external images.

I've also tried caching the images as local tar files, but there was no major additional time saving there compared to the dependency proxy.

**Comment by ayalon** [comment]
Could you share how you have done this? I'm highly interested.

**Comment by AkibaAT** [comment]
I've created a gist with a simplified recap: https://gist.github.com/AkibaAT/06ff24091055388a62813ff60368176b

The basic process is:
- ddev step creates a checksum of all files involved in creating the built ddev images, per default that's the web and the db services.
It checks in the GitLab registry if the image is already built, and if so, skips the build, and if not, builds the images and then pushes them to the registry.
It also create 2 override YAML files for docker compose. One to change the image reference and remove the build instructions, and the other to make the web service startable independently (for simple tasks like npm & composer install that don't need the full DDEV).
- Because the ddev gitlab ci image is based on the docker dind image, I start dockerd directly within it, instead of using an extra dind service.
- Since we store the override YAML as artifact of the .ddev directory, ddev automatically includes and uses it.
ddev currently always tries to docker compose build, but since we removed the build config via the override, that step finishes immediately.
- To make use of the GitLab dependency proxy, gitlab-preload-images script is called, which prefixes all images found in the generated compose-full yaml and downloads them via the proxy.
Since ddev expects them without a prefix, we also retag them to remove the proxy prefix.
- The second override YAML is stored outside the .ddev directory as we don't want ddev to include it automatically. It can instead be used to start the web service via docker compose directly.

**Comment by AkibaAT** [comment]
> 👍 my intention was to use `ddev debug download-images` to download the images and ship this image with the images pre-downloaded.

I've realized this idea via a GitHub Workflow now in my fork: https://github.com/AkibaAT/ddev-gitlab-ci/pkgs/container/ddev-gitlab-ci

It doesn't run download-images, because that also pulls in mariadb. Instead, I just pulled in all the images that are always required:
- busybox:latest
- ddev/ddev-ssh-agent:${DDEV_VERSION}
- ddev/ddev-traefik-router:${DDEV_VERSION}
- ddev/ddev-utilities:latest
- ddev/ddev-webserver:${DDEV_VERSION}

I've also adapted the image to start dockerd itself in the entrypoint, instead of having to start an extra dind service. Just run the image, and you're good to go to use ddev.

The GitHub Workflow itself is adapted as well. Instead of using QEMU, I switched the workflow to use the native ARM64 runner that's become available for public use a few months ago.

**Comment by rfay** [comment]
Sounds great! I've always found that the extraction from cache (in github actions) took longer than just downloading the images. I'm interested in what you did, but am not finding it (and am confused between using github and gitlab here)

**Comment by AkibaAT** [comment]
> Sounds great! I've always found that the extraction from cache (in github actions) took longer than just downloading the images. I'm interested in what you did, but am not finding it (and am confused between using github and gitlab here)

https://github.com/ddev/ddev-gitlab-ci/compare/main...AkibaAT:ddev-gitlab-ci:main

The image is being built via GitHub, like in the source. However, the current experimental builder can run privileged commands during the build process, which allows starting a dind daemon, and pulling in the images in the Dockerfile.

Before this new build option, the only way to create a preloaded image would have been to manually start it, run the commands, and then export the container as a new image.

**Comment by rfay** [comment]
Thanks, I thought you were caching the images themselves by exporting to cache. I see. you're pulling them and saving them away.

**Comment by AkibaAT** [comment]
Yeah, they're ready to go when the the image starts up, without having to restore them from a TAR or such. I had tried that way before, and it was slower than pulling them in from the GitLab proxy.

**Comment by rfay** [comment]
Pretty amazing that you build the gitlab stuff on github :)

**Comment by ayalon** [comment]
@AkibaAT What about using kaniko to build and push the docker images?
https://github.com/GoogleContainerTools/kaniko

**Comment by AkibaAT** [comment]
> [@AkibaAT](https://github.com/AkibaAT) What about using kaniko to build and push the docker images? https://github.com/GoogleContainerTools/kaniko

While that looks like a cool project I hadn't heard of before, I'm not sure how it could apply here. As far as I can tell, it wouldn't enable pulling in other images into the built docker image, and it seems to be about solving a problem we don't have in the context of building images on GitHub.

**Comment by ayalon** [comment]
@AkibaAT I have tried to use your image with the example from your fork in a normal GitLab Pipeline runner, but it fails with the error:
"Can't start dockerd as this container is not in privileged mode".

What is needed in GitLab to use the prebuild image? I have successfully used Docker in Docker (dind) which works and is set up in the runner.

I also tried to follow your gist and prebuild the images in GitLab CI, but I failed so far. It is highly complex.

My build process involves 2 main steps:

- Build step: Here I don't need a full DDEV, but I need `composer install `and `npm run build`. The artifacts are carried to the test step
- Test step: Here I need a DDEV instance with database (optionally preloaded with an SQL dump) to run some phpunit, E2E test

Until now, I used a custom Docker image to fulfill the build step, but it would be nice to also benefit here from DDEV because it has everything I need to build.

In the test step, I used the Image from [here](https://github.com/ddev/ddev-gitlab-ci/blob/main/docs%2Fdocker.md), but downloading all layers is very slow and resource intensive.

I tried to prebuild the images with your Gist, but I did not manage yet. The latest version of DDEV includes many images like xhgui I don't need and I did not find a way to omit them during the build step. I'm also using the elasticsearch addon but during the tests I don't want the whole elasticsearch container to be downloaded and started.

Maybe you can guide me a bit how to proceed in Gitlab CI pipeline?

**Comment by noahsiegrist** [comment]
The `FF_DISABLE_UMASK_FOR_DOCKER_EXECUTOR: 1` tag, needed for DDEV to write to `docroot/sites/default`, causes issues with the container's entry point in GitLab CI. To reproduce:

```bash
docker run --rm -it ghcr.io/akibaat/ddev-gitlab-ci:stable id -u
```
The entry point tries to start the Docker daemon. The gitlab-start-docker script works without privileged mode, suggesting we may not need it. Removing the echo command in the entry point might fix the fatal error caused by the flag.

**Proposed Solutions**

- Remove echo Command: Adjust the entry point to avoid the fatal error.
- Re-evaluate Privileged Mode: Confirm if the Docker daemon needs privileged mode.
- Alternative UMASK Handling: Find ways to manage UMASK for DDEV without the flag.
- Modify Entry Point: Ensure compatibility with `FF_DISABLE_UMASK_FOR_DOCKER_EXECUTOR: 1.`

**Questions**

- Any workarounds for DDEV in GitLab CI without this flag?
- Is privileged mode necessary, or can we rely on gitlab-start-docker?

Looking forward to your suggestions!

**Comment by AkibaAT** [comment]
Thanks for the interest and the tries. I suspect this attempt might work with less friction in deployment, if I instead built a docker dind image that comes preloaded with the always required ddev images. Going that route would work as drop-in replacement without needing any change to gitlab configs, apart from replacing the dind image.

Then again, having it all in one single images is something I want at least for my own use.

**Comment by AkibaAT** [comment]
> The `FF_DISABLE_UMASK_FOR_DOCKER_EXECUTOR: 1` tag, needed for DDEV to write to `docroot/sites/default`, causes issues with the container's entry point in GitLab CI. To reproduce:
> 
> docker run --rm -it ghcr.io/akibaat/ddev-gitlab-ci:stable id -u
> The entry point tries to start the Docker daemon. The gitlab-start-docker script works without privileged mode, suggesting we may not need it. Removing the echo command in the entry point might fix the fatal error caused by the flag.
> 
> **Proposed Solutions**
> 
> * Remove echo Command: Adjust the entry point to avoid the fatal error.
> * Re-evaluate Privileged Mode: Confirm if the Docker daemon needs privileged mode.
> * Alternative UMASK Handling: Find ways to manage UMASK for DDEV without the flag.
> * Modify Entry Point: Ensure compatibility with `FF_DISABLE_UMASK_FOR_DOCKER_EXECUTOR: 1.`
> 
> **Questions**
> 
> * Any workarounds for DDEV in GitLab CI without this flag?
> * Is privileged mode necessary, or can we rely on gitlab-start-docker?
> 
> Looking forward to your suggestions!

Privileged mode is necessary for dockerd. This applies to regular, as well as rootless versions.

However, I've run into a wholly different issue with the idea of baking in the images, that being the overlay used by docker. During build, the best available is fuse-overlayfs, and I've run into weird issues with that one when trying to also pull in the mariadb image, and then start the built image afterwards: `docker: Error response from daemon: failed to get file info for mysql.sock: lstat /var/lib/docker/rootfs/overlayfs/59e94b62794d0a31bdbcea878acab2c45972bc303d289e7120d821520e2f689d/var/lib/docker/fuse-overlayfs/dc338dc7885692f63d04b4e1345e92ed8cbff25c5f1efab236bf39c18c74b8e0/diff/var/tmp/mysql.sock: no such file or directory`
Additionally, when loading the image on GitLab, it defaults to overlay2, which means that none of the loaded images are being found, unless the storage is configured down to also use fuse-overlayfs.

So, I think I'll completely abandon the idea of baked in images, and instead concentrate on using the GitLab dependency proxy.

**Comment by AkibaAT** [comment]
Some testruns on GitLab.com:

- Simple DDEV project, no customization:
  - DDEV Image: About 2:30 every run
  - AkibaAT Image: About 5:30 the first run, then about 2:00 every following run
- Simple DDEV project with additional packages in web container:
  - DDEV Image: About 3:00 every run
  - AkibaAT Image: About 3:00 the first run (DDEV Images already proxied from previous testcase), then about 2:00 every following run

Not a huge difference on an installation with plenty available bandwidth. My version only really has a noticeable advantage if a project relies on heavy customization of the images, or is being run on a self-hosted GitLab instance with limited bandwidth. And it removes the need for Mutagen in the pipeline, an additional bonus to me.

**Comment by AkibaAT** [comment]
Testing it with the default DDEV image, I'm not sure why the documentation says `ddev config global --no-bind-mounts=true` is required. The pipeline seems just fine without it.

**Comment by rfay** [comment]
>  I'm not sure why the documentation says ddev config global --no-bind-mounts=true is required. The pipeline seems just fine without it.

In the past, you couldn't bind-mount to the Docker provider.  Maybe different Gitlab configurations?

**Comment by stasadev** [comment]
I saw an error without `--no-bind-mounts=true` here when I tried to remove it:

- https://github.com/ddev/ddev-gitlab-ci/pull/17#issuecomment-2721597907

**Comment by AkibaAT** [comment]
> I saw an error without `--no-bind-mounts=true` here when I tried to remove it:
> 
> * [fix: exclude v1.22 and v1.23 from builds because of Mutagen and Docker 28 error #17 (comment)]

Seems fine on my current GitLab.com run without it: https://gitlab.com/ddev-testing/ddev-ci-test/-/jobs/9774214049

**Comment by stasadev** [comment]
> Seems fine on my current GitLab.com run without it: https://gitlab.com/ddev-testing/ddev-ci-test/-/jobs/9774214049

It doesn't work with tests on GitHub in this repository. This is probably the only reason.

> Testing it with the default DDEV image, I'm not sure why the documentation says `ddev config global --no-bind-mounts=true` is required. The pipeline seems just fine without it.

Then it should be updated.

**Comment by ochorocho** [comment]
Yes, would be great if the `--no-bind-mounts=true` can be removed now.

**Comment by plan2github** [comment]
> It doesn't work with tests on GitHub in this repository. This is probably the only reason.

That makes sense. The way the GitHub Workflow is set up, there's no shared directory between the DIND and the DDEV images, which is why the bind mounts are failing. That would be fixable, then the `--no-bind-mounts=true` could be dropped from the GitHub Workflow as well.

**Comment by GaryWilsonJr** [comment]
> It doesn't work with tests on GitHub in this repository. This is probably the only reason.

That makes sense. The way the GitHub Workflow is set up, there's no shared directory between the DIND and the DDEV images, which is why the bind mounts are failing. That would be fixable, then the `--no-bind-mounts=true` could be dropped from the GitHub Workflow as well.

**Comment by ayalon** [comment]
I was not satified with the slowness of DIND and tried a different approach. It's not suitable for simple test runs but in larger environments it is quite helpful.

All you need is a VPS where you can install `ddev`, `docker `and a `gitlab runner`. Instead of doing DIND and docker, I was using a shell executor. In `before_script` and `after_script` I do a `ddev stop` .

Additionally, some small overrides in `.ddev/config.local.yaml` is needed to provide a custom project name.
If you have the possibility to run a gitlab runner on a VPS you will gain an amazing pipeline speed. Docker images are cached and thanks to the global shared cache of DDEV also composer and npm is blazing fast.

**Comment by AkibaAT** [comment]
You could reach the same as you're describing by just using a runner mounted docker.sock. The volumes and images would be persisted as well then, giving you the identical up- and downsides, without needing to set up an additional VPS. At least on a self-hosted installation, not sure how hosted Gitlab.com handles persistence, if at all.

**Comment by ayalon** [comment]
What do you mean with mounted docker.sock? Instead of using a shell executor using a docker executor with a mounted docker.sock on a VPS?

**Comment by AkibaAT** [comment]
Yes. The slowness of the DIND solution stems from there being no persistence Docker side.
If you have a dockerd running on the host, and set up a regular docker runner, then all the images would stay cached on the host, as, unlike the dind container, its contents aren't forgotten with the end of the run.

**Comment by AkibaAT** [comment]
Alternatively, if you'd like to isolate the GitLab runs from other Docker services that might run on the same machine, you could configure a DIND service in the runner config, and add a volume mapping to it. Then the DIND data would be persisted, with all the up- and downsides that brings.

Sort of like this (written from memory, might be a bit off):
```
[[runners]]
  [runners.docker]
    volumes = [
      "/var/lib/docker"
    ]
  [[runners.docker.services]]
    name = "docker:dind"
```

**Comment by AkibaAT** [comment]
> > Seems fine on my current GitLab.com run without it: https://gitlab.com/ddev-testing/ddev-ci-test/-/jobs/9774214049
> 
> It doesn't work with tests on GitHub in this repository. This is probably the only reason.
> 
> > Testing it with the default DDEV image, I'm not sure why the documentation says `ddev config global --no-bind-mounts=true` is required. The pipeline seems just fine without it.
> 
> Then it should be updated.

Opened a PR to remove no-bind-mounts, including from the GitHub Workflow: https://github.com/ddev/ddev-gitlab-ci/pull/23
